packages/truffle-decoder/lib/decode/value.ts

```diff
@@ -3,177 +3,364 @@ const debug = debugModule("decoder:decode:value");
+    );
   }
   //also before we continue: is the PC zero? if so let's just return that
-  if(pc === 0) {
-    return "<zero>";
+  if(deployedPc === 0 && constructorPc === 0) {
+    return new Values.FunctionInternalValue(
+      dataType,
+      new Values.FunctionInternalValueInfoException(context, deployedPc, constructorPc)
+    );
+  }
+  //another check: is only the deployed PC zero?
+  if(deployedPc === 0 && constructorPc !== 0) {
+    return new Values.FunctionInternalErrorResult(
+      dataType,
+      new Values.MalformedInternalFunctionError(context, constructorPc)
+    );
+  }
+  //one last pre-check: is this a deployed-format pointer in a constructor?
+  if(info.currentContext.isConstructor && constructorPc === 0) {
+    return new Values.FunctionInternalErrorResult(
+      dataType,
+      new Values.DeployedFunctionInConstructorError(context, deployedPc)
+    );
   }
   //otherwise, we get our function
+  let pc = info.currentContext.isConstructor
+    ? constructorPc
+    : deployedPc;
   let functionEntry = info.internalFunctionsTable[pc];
   if(!functionEntry) {
-    //if it's not zero and there's no entry... we give up :P
-    return undefined;
+    //if it's not zero and there's no entry... error!
+    return new Values.FunctionInternalErrorResult(
+      dataType,
+      new Values.NoSuchInternalFunctionError(context, deployedPc, constructorPc)
+    );
   }
   if(functionEntry.isDesignatedInvalid) {
-    return "assert(false)";
+    return new Values.FunctionInternalValue(
+      dataType,
+      new Values.FunctionInternalValueInfoException(context, deployedPc, constructorPc)
+    );
   }
-  return functionEntry.contractName + "." + functionEntry.name;
+  let name = functionEntry.name;
+  let definedIn: Types.ContractType = {
+    typeClass: "contract",
+    id: functionEntry.contractId,
+    typeName: functionEntry.contractName,
+    contractKind: functionEntry.contractKind,
+    payable: functionEntry.contractPayable
+  };
+  return new Values.FunctionInternalValue(
+    dataType,
+    new Values.FunctionInternalValueInfoKnown(context, deployedPc, constructorPc, name, definedIn)
+  );
+}
+
+function checkPaddingRight(bytes: Uint8Array, length: number): boolean {
+  let padding = bytes.slice(length); //cut off the first length bytes
+  return padding.every(paddingByte => paddingByte === 0);
+}
+
+//exporting this one for use in stack.ts
+export function checkPaddingLeft(bytes: Uint8Array, length: number): boolean {
+  let padding = bytes.slice(0, -length); //cut off the last length bytes
+  return padding.every(paddingByte => paddingByte === 0);
+}
+
+function checkPaddingSigned(bytes: Uint8Array, length: number): boolean {
+  let padding = bytes.slice(0, -length); //padding is all but the last length bytes
+  let value = bytes.slice(-length); //meanwhile the actual value is those last length bytes
+  let signByte = value[0] & 0x80 ? 0xff : 0x00;
+  return padding.every(paddingByte => paddingByte === signByte);
 }
```

Comment:
What do you think? Not too bad, right? Seems to follow. Notes about cohesion from above probably still apply here, but these `lib/decode` and `lib/allocate` modules really seem to follow nicely from the all the types and values setup. Cool!

Reply:
Sorry, what is this referring to?  It looks like this got misplaced.

Reply:
oh I just mean the ramifications of types.ts and values.ts are "not too bad". Like, `lib/decode` and `lib/allocate` are moving in the right direction, I feel.